src/storage/mod.rs

```diff
@@ -191,6 +191,96 @@ where
     }
 }
 
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> &T {
+        unsafe { self.storage.data.inner.get(self.entity.id()) }
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(self) -> &'a mut T {
+        unsafe { self.storage.data.inner.get_mut(self.entity.id()) }
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    pub fn insert(&mut self, mut component: T) -> T {
+        std::mem::swap(&mut component, unsafe { self.storage.data.inner.get_mut(self.entity.id()) });
```

Comment:
could use `self.get_mut()`

Reply:
Annoyingly I can't figure out the lifetimes that would work for `get_mut` that allows just `&mut self`, so I can't get that to work properly.